network/simple-onchain-discovery/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+            "Histogram of idle time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+/// Histogram of busy time of spent in event processing loop
+pub static EVENT_PROCESSING_LOOP_BUSY_DURATION_S: Lazy<DurationHistogram> = Lazy::new(|| {
+    DurationHistogram::new(
+        register_histogram!(
+            "simple_onchain_discovery_event_processing_loop_busy_duration_s",
+            "Histogram of busy time of spent in event processing loop"
+        )
+        .unwrap(),
+    )
+});
+
+/// Listener which converts published  updates from the OnChainConfig to ConnectivityRequests
+/// for the ConnectivityManager.
+pub struct ConfigurationChangeListener {
+    conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
+    role: RoleType,
+}
+
+/// Extracts a set of ConnectivityRequests from a ValidatorSet which are appropriate for a Validator Network.
+fn extract_validator_updates(validator_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let validator_keys = validator_set.payload().to_vec();
+
+    let mut updates = Vec::new();
+
+    // Collect the set of address updates.
+    let address_map = validator_keys
+        .iter()
+        .map(|validator| {
+            (
+                validator.account_address().clone(),
+                vec![
+                    NetworkAddress::try_from(&validator.config().validator_network_address)
+                        .expect("WTF"),
+                ],
+            )
+        })
+        .collect();
+
+    let update_address_req =
+        ConnectivityRequest::UpdateAddresses(DiscoverySource::OnChain, address_map);
+
+    updates.push(update_address_req);
+
+    // Collect the set of EligibleNodes
+    updates.push(ConnectivityRequest::UpdateEligibleNodes(
+        validator_keys
+            .into_iter()
+            .map(|validator| {
+                (
+                    *validator.account_address(),
+                    NetworkPublicKeys {
+                        identity_public_key: validator.network_identity_public_key(),
+                        signing_public_key: validator.network_signing_public_key().clone(),
+                    },
+                )
+            })
+            .collect(),
+    ));
+
+    updates
+}
+
+/// Extracts a set of ConnectivityRequests from a ValidatorSet which are appropriate for a FullNode Network.
+fn extract_full_node_updates(full_node_set: ValidatorSet) -> Vec<ConnectivityRequest> {
+    let full_node_infos = full_node_set.payload().to_vec();
+
+    let mut updates = Vec::new();
+
+    // Collect the set of updated addresses.
+    let address_map = full_node_infos
+        .iter()
+        .map(|full_node| {
+            (
+                full_node.account_address().clone(),
+                vec![NetworkAddress::try_from(
+                    &full_node.config().full_node_network_address.clone(),
+                )
+                .expect("WTF")],
```

Comment:
same comment as above: let's log an error here and not panic.... someone could accidentally put bad data here, this shouldn't kill validators or other nodes.

Reply:
Done.